src/derive_utils.rs

```diff
@@ -66,33 +68,36 @@ impl FunctionDescription {
         );
 
         // Handle positional arguments
-        let (args_provided, varargs) = {
+        let args_provided = {
             let args_provided = args.len();
-
             if self.accept_varargs {
-                (
-                    std::cmp::min(num_positional_parameters, args_provided),
-                    Some(args.slice(num_positional_parameters as isize, args_provided as isize)),
-                )
+                std::cmp::min(num_positional_parameters, args_provided)
             } else if args_provided > num_positional_parameters {
                 return Err(self.too_many_positional_arguments(args_provided));
             } else {
-                (args_provided, None)
+                args_provided
             }
         };
 
         // Copy positional arguments into output
-        for (out, arg) in output[..args_provided].iter_mut().zip(args) {
+        for (out, arg) in output[..args_provided].iter_mut().zip(args.by_ref()) {
             *out = Some(arg);
         }
 
+        // Collect varargs into tuple
+        let varargs = if self.accept_varargs {
+            Some(PyTuple::new(py, args))
```

Comment:
Hm, what would you optimize there?

Reply:
I see two possible inefficiencies in `PyTuple::new` which we could improve:
  - It uses `PyTuple_SetItem` but could use `PyTuple_SET_ITEM`, which does less error checking. I think we would also have to define `PyTuple_SET_ITEM` in Rust (it's a C macro), so this would inline really well.
  - It uses `.to_object().into_ptr()` for generic `T`, if we wrote an implementation of `PyTuple::new` which took `&PyAny` iterator we could just use `into_ptr()` which might make a teeeny tiny performance improvement (if the optimizer didn't already achieve the same effect).

I definitely think the first bullet could be worth trying. Could make a nice standalone PR rather than being part of this one.